lib/plenario_web/controllers/api/aot_controller.ex

```diff
@@ -1,236 +1,108 @@
-    obs_data_ids =
-      case Map.get(conn.params, "sensor") do
-        nil ->
-          nil
-
-        value ->
-          data_ids = Repo.all(
-            from d in data,
-            select: d.id,
-            distinct: d.id
-          )
-          case is_list(value) do
-            true ->
-              Repo.all(
-                from o in AotObservation,
-                where: o.aot_data_id in ^data_ids,
-                where: o.sensor in ^value,
-                select: o.aot_data_id,
-                distinct: o.aot_data_id
-              )
-            false ->
-              Repo.all(
-                from o in AotObservation,
-                where: o.aot_data_id in ^data_ids,
-                where: o.sensor == ^value,
-                select: o.aot_data_id,
-                distinct: o.aot_data_id
-              )
-          end
-      end
-
-    # if observation limited query, apply to data query
-    data =
-      case obs_data_ids do
-        nil -> data
-        value -> from d in data, where: d.id in ^value
-      end
-
-    # finally apply the windowing `inserted_at` filter
-    window =
-      case Map.get(conn.params, "inserted_at") do
-        nil -> DateTime.utc_now()
-        value ->
-          [_, term] = String.split(value, ":", parts: 2)
-          parse_datetime(term)
-      end
-    data = from d in data, where: d.inserted_at <= ^window
-
-    # return meta and data queries
-    [meta_query: metas, data_query: data]
-  end
-
-  def get(conn, _) do
-    [meta_query: _, data_query: query] = handle_conn_params(conn)
-
-    pagination = [
-      page: Map.get(conn.params, "page", 1),
-      page_size: Map.get(conn.params, "page_size", 500)
-    ]
-
-    page = Repo.paginate(query, pagination)
-    render_page(conn, "get.json", conn.params, page.entries, page, true)
-  end
+  defp get_data({:error, message}, _), do: {:error, message}
 
-  def head(conn, _) do
-    [meta_query: _, data_query: query] = handle_conn_params(conn)
+  defp get_data({:ok, meta_ids}, conn) do
+    query =
+      conn.assigns[:filters]
+      |> Enum.reject(fn {fname, _, _} -> fname in @meta_keys end)
+      |> Enum.reduce(AotData, fn {fname, op, value}, query ->
+        apply_filter(query, fname, op, value)
+      end)
+      |> where([q], q.aot_meta_id in ^meta_ids)
+      |> where([q], q.updated_at <= ^conn.assigns[:window])
+      |> join(:left, [d], m in assoc(d, :aot_meta))
+      |> select([d, m], %{d | aot_meta: %{network_name: m.network_name, slug: m.slug}})
 
-    page = Repo.paginate(query, page: 1, page_size: 1)
-    render_page(conn, "get.json", conn.params, page.entries, page, true)
+    try do
+      page = conn.assigns[:page]
+      page_size = conn.assigns[:page_size]
+      data = Repo.paginate(query, page: page, page_size: page_size)
+      {:ok, data}
+    rescue
+      e in [Ecto.QueryError, Ecto.SubQueryError, Postgrex.Error] ->
+        {:error, e.message}
+    end
```

Comment:
Handling errors like this makes it tough to distinguish between what is bad user input and what is a true error on our part. At the very least, `halt_with` should send a copy of the error to sentry, or this head ` defp render_data({:error, message}, conn, _)` of `render_data` should send information to sentry. Sweeping `try-rescues` also runs counter to the Erlang "let it die" philosophy. Bad input shouldn't cause us to crash, bad input we handle incorrectly should.

Reply:
The reason for this is because I got rid of the explicit field naming in the plugs -- if a user passes a param `foo=bar` and there isn't a `foo` field on the data set, it will error.

This isn't an uncaught internal error on our part, this is a client error. I catch it and send it back to them.